alf/utils/distributions.py

```diff
@@ -0,0 +1,318 @@
+
+
+class TruncatedDistribution(td.Distribution):
+    r"""The base class of truncated distributions.
+
+    A truncated distribution :math:`q(x)` is defined as a standard base distribution :math:`p(x)` and
+    location :math:`\mu`, scale parameters :math:`s`, lower bound :math:`l` and
+    upper bound :math:`u`
+
+    .. math::
+
+        q(x) = \frac{1}{s (P(u) - P(l))}p(\frac{x-\mu}{s})
+
+    where :math:`P` is the cdf of :math:`p`.
+
+    Args:
+        loc: the location parameter
+        scale: the scale parameter
+        lower_bound: the lower bound
+        upper_bound: the upper bound
+        its: the standard distribution to be used.
+    """
+
+    arg_constraints = {
+        'loc': td.constraints.real,
+        'scale': td.constraints.positive
+    }
+    has_rsample = True
+
+    def __init__(self, loc: Tensor, scale: Tensor, lower_bound: Tensor,
+                 upper_bound: Tensor, its: InverseTransformSampling):
+        event_shape = torch.broadcast_shapes(lower_bound.shape,
+                                             upper_bound.shape)
+        batch_shape = torch.broadcast_shapes(scale.shape, loc.shape,
+                                             event_shape)
+        if len(event_shape) > 0:
+            batch_shape = batch_shape[:-len(event_shape)]
+
+        self._scale = scale
+        self._loc = loc
+
+        super().__init__(batch_shape=batch_shape, event_shape=event_shape)
+        self._its = its
+        self._lower_bound = lower_bound
+        self._upper_bound = upper_bound
+        self._cdf_lb = self._its.cdf((lower_bound - loc) / scale)
+        self._cdf_ub = self._its.cdf((upper_bound - loc) / scale)
+        self._logz = (scale * (self._cdf_ub - self._cdf_lb + 1e-30)).log()
+
+    @property
+    def scale(self):
+        """Scale parameter of this distribution."""
+        return self._scale
+
+    @property
+    def loc(self):
+        """Location parameter of this distribution."""
+        return self._loc
+
+    @property
+    def lower_bound(self):
+        """Lower bound of this distribution."""
+        return self._lower_bound
+
+    @property
+    def upper_bound(self):
+        """Upper bound of this distribution."""
+        return self._upper_bound
+
+    @property
+    def mode(self):
+        """Mode of this distribution."""
+        return self._loc.clamp(self._lower_bound, self._upper_bound)
+
+    def rsample(self, sample_shape: torch.Size = torch.Size()):
+        """
+        Generates a sample_shape shaped reparameterized sample or sample_shape
+        shaped batch of reparameterized samples if the distribution parameters
+        are batched.
+
+        Args:
+            sample_shape: sample shape
+        Returns:
+            Tensor of shape ``sample_shape + batch_shape + event_shape``
+        """
+        r = torch.rand(sample_shape + self._batch_shape + self._event_shape)
+        r = (1 - r) * self._cdf_lb + r * self._cdf_ub
+        r = r.clamp(0.001, 0.999)
```

Comment:
Is 1e-3 the standard way of doing this clipping? It seems a somewhat arbitrary decision.